File: app/src/main/java/com/aaron/yespdf/filepicker/ScanActivity.java
package com.aaron.yespdf.filepicker;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.SystemClock;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.aaron.base.impl.OnClickListenerImpl;
import com.aaron.base.impl.TextWatcherImpl;
import com.aaron.base.util.StatusBarUtils;
import com.aaron.yespdf.R;
import com.aaron.yespdf.R2;
import com.aaron.yespdf.common.CommonActivity;
import com.aaron.yespdf.common.DialogManager;
import com.aaron.yespdf.common.GroupingAdapter;
import com.aaron.yespdf.common.UiManager;
import com.blankj.utilcode.util.KeyboardUtils;
import com.blankj.utilcode.util.LogUtils;
import com.blankj.utilcode.util.SDCardUtils;
import com.blankj.utilcode.util.StringUtils;
import com.github.anzewei.parallaxbacklayout.ParallaxBack;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.uber.autodispose.AutoDispose;
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@ParallaxBack
public class ScanActivity extends CommonActivity {

    static final String EXTRA_IMPORTED = "EXTRA_IMPORTED";

    @BindView(R2.id.app_search_view)
    View searchView;
    @BindView(R2.id.app_ibtn_back)
    ImageButton ibtnCancelSearch;
    @BindView(R2.id.app_et_search)
    EditText etSearch;
    @BindView(R2.id.app_ibtn_inverse)
    ImageButton ibtnInverse;
    @BindView(R2.id.app_ibtn_clear)
    ImageButton ibtnClear;
    @BindView(R2.id.app_ibtn_search)
    ImageButton ibtnSearch;
    @BindView(R2.id.app_ibtn_select_all)
    ImageButton ibtnSelectAll;
    @BindView(R2.id.app_rv_select)
    RecyclerView rv;
    @BindView(R2.id.app_btn_import_count)
    Button btnImport;

    // 揭露动画参数
    private int duration = 250;
    private int centerX;
    private int centerY;
    private float radius;

    private List<File> fileList = new ArrayList<>();
    private List<String> selectList = new ArrayList<>();

    private TextView tvScanCount;
    private TextView tvPdfCount;
    private Button btnStopScan;
    private int scanCount;
    private int pdfCount;
    private boolean stopScan;
    private String newGroupName;

    private ExecutorService threadPool;

    private Unbinder unbinder;
    private ViewAllAdapter adapter;
    private BottomSheetDialog scanDialog;
    private BottomSheetDialog importDialog;
    private BottomSheetDialog groupingDialog;
    private IListable listable = new ByNameListable();
    private RecyclerView.AdapterDataObserver dataObserver = new RecyclerView.AdapterDataObserver() {
        @Override
        public void onChanged() {
            if (!isFinishing()) {
                selectList.clear();
                ibtnSelectAll.setSelected(false);
                btnImport.setText(R.string.app_import_count);
                if (stopScan) {
                    boolean enableSelectAll = adapter.reset();
                    ibtnSelectAll.setEnabled(enableSelectAll);
                }
            }
        }
    };

    public static void start(Activity activity, ArrayList<String> imported, int requestCode) {
        Intent starter = new Intent(activity, ScanActivity.class);
        starter.putStringArrayListExtra(EXTRA_IMPORTED, imported);
        activity.startActivityForResult(starter, requestCode);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        unbinder = ButterKnife.bind(this);
        initToolbar();
        toolbar.setPadding(0, 0, 0, 0);
        StatusBarUtils.setStatusBarLight(this, true);
        initView();
    }

    @Override
    protected int layoutId() {
        return R.layout.app_activity_scan;
    }

    @Override
    protected Toolbar createToolbar() {
        return findViewById(R.id.app_toolbar);
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (KeyboardUtils.isSoftInputVisible(this)) {
            KeyboardUtils.hideSoftInput(this);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbinder.unbind();
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }

    @Override
    public void onBackPressed() {
        if (searchView.getVisibility() == View.VISIBLE) {
            closeSearchView();
        } else {
            super.onBackPressed();
        }
    }

    @SuppressLint("SetTextI18n")
    private void initView() {
        List<String> imported = getIntent().getStringArrayListExtra(EXTRA_IMPORTED);

        threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());

        scanDialog = DialogManager.createScanDialog(this, new DialogManager.ScanDialogCallback() {
            @Override
            public void onTitle(TextView tv) {
                tvScanCount = tv;
                tvScanCount.setText(getString(R.string.app_already_scan) + 0 + getString(R.string.app_file_));
            }

            @Override
            public void onContent(TextView tv) {
                tvPdfCount = tv;
                tvPdfCount.setText(getString(R.string.app_find) + "PDF(" + 0 + ")");
            }

            @Override
            public void onButton(Button btn) {
                btnStopScan = btn;
                btnStopScan.setOnClickListener(new OnClickListenerImpl() {
                    @Override
                    public void onViewClick(View v, long interval) {
                        btnStopScan.setSelected(true);
                        stopScan = true;
                        scanDialog.dismiss();
                        threadPool.shutdownNow();
                        updateUI();
                    }
                });
            }
        });

        searchView.post(() -> {
            centerX = ibtnSearch.getLeft() + ibtnSearch.getMeasuredWidth() / 2;
            centerY = ibtnSearch.getTop() + ibtnSearch.getMeasuredHeight() / 2;
            int width = centerX * 2;
            int height = searchView.getMeasuredHeight();
            radius = (float) (Math.sqrt(width * width + height * height) / 2);
        });

        ibtnCancelSearch.setOnClickListener(v -> {
            closeSearchView();
        });
        ibtnSearch.setOnClickListener(v -> openSearchView());
        ibtnInverse.setOnClickListener(v -> {
            if (KeyboardUtils.isSoftInputVisible(this)) {
                KeyboardUtils.hideSoftInput(this);
            }
            ibtnInverse.setSelected(!ibtnInverse.isSelected());
            adapter.setInverse(ibtnInverse.isSelected());
            adapter.getFilter().filter(etSearch.getText());
        });
        ibtnClear.setOnClickListener(v -> {
            etSearch.setText("");
            if (!KeyboardUtils.isSoftInputVisible(this)) {
                KeyboardUtils.showSoftInput(this);
            }
        });
        etSearch.addTextChangedListener(new TextWatcherImpl() {
            @Override
            public void onTextChanged(CharSequence c, int i, int i1, int i2) {
                ibtnInverse.setVisibility(c.length() == 0 ? View.GONE : View.VISIBLE);
                ibtnClear.setVisibility(c.length() == 0 ? View.GONE : View.VISIBLE);
                adapter.getFilter().filter(c);
            }
        });
        ibtnSelectAll.setOnClickListener(v -> {
            v.setSelected(!v.isSelected());
            adapter.selectAll(v.isSelected());
        });
        btnImport.setOnClickListener(new OnClickListenerImpl() {
            @Override
            public void onViewClick(View v, long interval) {
                if (KeyboardUtils.isSoftInputVisible(ScanActivity.this)) {
                    KeyboardUtils.hideSoftInput(ScanActivity.this);
                }
                if (selectList.isEmpty()) {
                    UiManager.showShort(R.string.app_have_not_select);
                } else {
                    if (importDialog == null) {
                        initImportDialog();
                    }
                    importDialog.show();
                }
            }
        });

        ibtnSelectAll.setEnabled(false); // XML 设置无效，只能这里初始化
        RecyclerView.LayoutManager lm = new LinearLayoutManager(this);
        rv.setLayoutManager(lm);
        adapter = new ViewAllAdapter(fileList, imported, new ViewAllAdapter.Callback() {
            @Override
            public void onDirTap(String dirPath) {
                // empty impl
            }

            @SuppressLint("SetTextI18n")
            @Override
            public void onSelectResult(List<String> pathList, int total) {
                LogUtils.e(pathList.size());
                if (total != 0) {
                    ibtnSelectAll.setSelected(pathList.size() == total);
                }
                btnImport.setText(getString(R.string.app_import) + "(" + pathList.size() + ")");
                selectList.clear();
                selectList.addAll(pathList);
            }
        });
        adapter.registerAdapterDataObserver(dataObserver);
        rv.setAdapter(adapter);

        scanDialog.show();
        Observable.create((ObservableOnSubscribe<Double>) emitter -> {
                    emitter.onNext(traverseFile());
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .as(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(this)))
                .subscribe(cost -> {
                    if (!btnStopScan.isSelected()) {
                        scanDialog.dismiss();
                        updateUI();
                    }
                }, throwable -> LogUtils.e(throwable.getMessage()));
    }

    private double traverseFile() {
        long start = System.currentTimeMillis();
        List<SDCardUtils.SDCardInfo> result = SDCardUtils.getSDCardInfo();
        for (SDCardUtils.SDCardInfo info : result) {
            if (!stopScan && "mounted".equals(info.getState())) {
                traverse(new File(info.getPath()));
            }
        }
        while (true) {
            int temp = scanCount;
            SystemClock.sleep(1000);
            if (temp == scanCount) {
                stopScan = true;
                long end = System.currentTimeMillis();
                double cost = (double) (end - start - 1000) / 1000;
                LogUtils.e("总共耗时：" + cost + " 秒");
                return cost;
            }
        }
    }

//    private int refreshLayoutFlag = 2000;

    @SuppressLint("SetTextI18n")
    private synchronized void traverse(File file) {
        if (stopScan) {
            return;
        }
        threadPool.execute(() -> {
            List<File> fileList = listable.listFile(file.getAbsolutePath());
            for (File f : fileList) {
                if (stopScan) {
                    return;
                }
                scanCount++;
                runOnUiThread(() -> {
                    tvScanCount.setText(getString(R.string.app_already_scan) + scanCount + getString(R.string.app_file_));
                });
                if (f.isFile()) {
                    this.fileList.add(f);
                    pdfCount++;
                    runOnUiThread(() -> {
                        toolbar.setTitle(getString(R.string.app_scan_result) + "(" + pdfCount + ")");
                        tvPdfCount.setText(getString(R.string.app_find) + "PDF(" + pdfCount + ")");
//                        if (scanCount > refreshLayoutFlag) {
//                            adapter.notifyDataSetChanged();
//                            refreshLayoutFlag += 5000;
//                        }
                    });
                } else {
                    traverse(f);
                }
            }
        });
    }

    private void initImportDialog() {
        importDialog = DialogManager.createImportDialog(this, new DialogManager.ImportDialogCallback() {
            @Override
            public void onInput(EditText et) {
                et.addTextChangedListener(new TextWatcherImpl() {
                    @Override
                    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                        newGroupName = charSequence.toString();
                    }
                });
            }

            @Override
            public void onLeft(Button btn) {
                btn.setOnClickListener(new OnClickListenerImpl() {
                    @Override
                    public void onViewClick(View v, long interval) {
                        if (groupingDialog == null) {
                            initGroupDialog();
                        }
                        groupingDialog.show();
                    }
                });
            }

            @Override
            public void onCenter(Button btn) {
                btn.setOnClickListener(new OnClickListenerImpl() {
                    @Override
                    public void onViewClick(View v, long interval) {
                        setResultBack(SelectActivity.TYPE_BASE_FOLDER, null);
                    }
                });
            }

            @Override
            public void onRight(Button btn) {
                btn.setOnClickListener(new OnClickListenerImpl() {
                    @Override
                    public void onViewClick(View v, long interval) {
                        if (StringUtils.isEmpty(newGroupName)) {
                            UiManager.showShort(R.string.app_type_new_group_name);
                        } else {
                            setResultBack(SelectActivity.TYPE_CUSTOM, newGroupName);
                        }
                    }
                });
            }
        });
    }

    private void initGroupDialog() {
        groupingDialog = DialogManager.createGroupingDialog(ScanActivity.this, false, new GroupingAdapter.Callback() {
            @Override
            public void onAddNewGroup() {
                // empty
            }

            @Override
            public void onAddToGroup(String dir) {
                setResultBack(SelectActivity.TYPE_TO_EXIST, dir);
            }
        });
    }

    private void initToolbar() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowTitleEnabled(false);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeAsUpIndicator(R.drawable.app_ic_action_back_black);
        }
        toolbar.setTitle(R.string.app_scan_result);
    }

    private void openSearchView() {
        Animator animator = ViewAnimationUtils.createCircularReveal(searchView, centerX, centerY, 0, radius);
        animator.setDuration(duration);
        animator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                searchView.setVisibility(View.VISIBLE);
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                etSearch.requestFocus();
                KeyboardUtils.showSoftInput(ScanActivity.this);
            }
        });
        animator.start();
    }

    private void closeSearchView() {
        Animator animator = ViewAnimationUtils.createCircularReveal(searchView, centerX, centerY, radius, 0);
        animator.setDuration(duration);
        animator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                if (KeyboardUtils.isSoftInputVisible(ScanActivity.this)) {
                    KeyboardUtils.hideSoftInput(ScanActivity.this);
                }
                searchView.setVisibility(View.GONE);
            }
        });
        animator.start();
    }

    private void setResultBack(int type, String groupName) {
        Intent data = new Intent();
        data.putStringArrayListExtra(SelectActivity.EXTRA_SELECTED, (ArrayList<String>) selectList);
        data.putExtra(SelectActivity.EXTRA_TYPE, type);
        data.putExtra(SelectActivity.EXTRA_GROUP_NAME, groupName);
        setResult(RESULT_OK, data);
        finish();
    }

    private void updateUI() {
        LayoutAnimationController lac = AnimationUtils.loadLayoutAnimation(this, R.anim.app_layout_fall_down);
        rv.setLayoutAnimation(lac);
        rv.getAdapter().notifyDataSetChanged();
        rv.scheduleLayoutAnimation();
    }
}

File: app/src/main/java/com/aaron/yespdf/filepicker/ScanActivity.kt
package com.aaron.yespdf.filepicker

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.SystemClock
import android.view.View
import android.view.ViewAnimationUtils
import android.view.animation.AnimationUtils
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.AdapterDataObserver
import com.aaron.base.impl.OnClickListenerImpl
import com.aaron.base.impl.TextWatcherImpl
import com.aaron.base.util.StatusBarUtils
import com.aaron.yespdf.R
import com.aaron.yespdf.common.CommonActivity
import com.aaron.yespdf.common.DialogManager
import com.aaron.yespdf.common.GroupingAdapter
import com.aaron.yespdf.common.UiManager
import com.aaron.yespdf.filepicker.ScanActivity
import com.blankj.utilcode.util.KeyboardUtils
import com.blankj.utilcode.util.LogUtils
import com.blankj.utilcode.util.SDCardUtils
import com.blankj.utilcode.util.StringUtils
import com.github.anzewei.parallaxbacklayout.ParallaxBack
import com.google.android.material.bottomsheet.BottomSheetDialog
import io.reactivex.Observable
import io.reactivex.ObservableEmitter
import io.reactivex.ObservableOnSubscribe
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.app_activity_scan.*
import kotlinx.android.synthetic.main.app_include_searchview.*
import java.io.File
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.math.sqrt

@ParallaxBack
class ScanActivity : CommonActivity() {

    // 揭露动画参数
    private val duration = 250
    private var centerX = 0
    private var centerY = 0
    private var radius = 0f
    private val fileList: MutableList<File> = ArrayList()
    private val selectList: MutableList<String> = ArrayList()
    private var tvScanCount: TextView? = null
    private var tvPdfCount: TextView? = null
    private var btnStopScan: Button? = null
    private var scanCount = 0
    private var pdfCount = 0
    private var stopScan = false
    private var newGroupName: String? = null
    private var threadPool: ExecutorService? = null
    private val listable: IListable = ByNameListable()

    private var adapter: ViewAllAdapter? = null

    private val scanDialog: BottomSheetDialog by lazy(LazyThreadSafetyMode.NONE) {
        DialogManager.createScanDialog(this) { tvTitle, tvContent, btn ->
            tvScanCount = tvTitle
            tvScanCount?.text = getString(R.string.app_already_scan, 0)
            tvPdfCount = tvContent
            tvPdfCount?.text = getString(R.string.app_find, 0)
            btnStopScan = btn
            btnStopScan?.setOnClickListener(object : OnClickListenerImpl() {
                override fun onViewClick(v: View, interval: Long) {
                    btnStopScan?.isSelected = true
                    stopScan = true
                    scanDialog.dismiss()
                    threadPool?.shutdownNow()
                    updateUI()
                }
            })
        }
    }
    private val importDialog: BottomSheetDialog by lazy(LazyThreadSafetyMode.NONE) {
        DialogManager.createImportDialog(this) { etInput, btnLeft, btnCenter, btnRight ->
            etInput.addTextChangedListener(object : TextWatcherImpl() {
                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                    newGroupName = charSequence.toString()
                }
            })
            btnLeft.setOnClickListener(object : OnClickListenerImpl() {
                override fun onViewClick(v: View, interval: Long) {
                    groupingDialog.show()
                }
            })
            btnCenter.setOnClickListener(object : OnClickListenerImpl() {
                override fun onViewClick(v: View, interval: Long) {
                    setResultBack(SelectActivity.TYPE_BASE_FOLDER, null)
                }
            })
            btnRight.setOnClickListener(object : OnClickListenerImpl() {
                override fun onViewClick(v: View, interval: Long) {
                    if (StringUtils.isEmpty(newGroupName)) {
                        UiManager.showShort(R.string.app_type_new_group_name)
                    } else {
                        setResultBack(SelectActivity.TYPE_CUSTOM, newGroupName)
                    }
                }
            })
        }
    }
    private val groupingDialog: BottomSheetDialog by lazy(LazyThreadSafetyMode.NONE) {
        DialogManager.createGroupingDialog(this@ScanActivity, false, object : GroupingAdapter.Callback {
            override fun onAddNewGroup() { // empty
            }

            override fun onAddToGroup(dir: String) {
                setResultBack(SelectActivity.TYPE_TO_EXIST, dir)
            }
        })
    }
    private val dataObserver: AdapterDataObserver = object : AdapterDataObserver() {
        override fun onChanged() {
            if (!isFinishing) {
                selectList.clear()
                app_ibtn_select_all.isSelected = false
                app_btn_import_count.setText(getString(R.string.app_import_count, 0))
                if (stopScan) {
                    val enableSelectAll = adapter!!.reset()
                    app_ibtn_select_all.isEnabled = enableSelectAll
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initToolbar()
        toolbar?.setPadding(0, 0, 0, 0)
        StatusBarUtils.setStatusBarLight(this, true)
        initView()
    }

    override fun layoutId(): Int {
        return R.layout.app_activity_scan
    }

    override fun createToolbar(): Toolbar? {
        return findViewById(R.id.app_toolbar)
    }

    override fun onStop() {
        super.onStop()
        if (KeyboardUtils.isSoftInputVisible(this)) {
            KeyboardUtils.hideSoftInput(this)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    override fun onBackPressed() {
        if (app_search_view.visibility == View.VISIBLE) {
            closeSearchView()
        } else {
            super.onBackPressed()
        }
    }

    @SuppressLint("SetTextI18n")
    private fun initView() {
        val imported: List<String>? = intent.getStringArrayListExtra(EXTRA_IMPORTED)
        threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors())
        app_search_view.post {
            centerX = app_ibtn_search.left + app_ibtn_search.measuredWidth / 2
            centerY = app_ibtn_search.top + app_ibtn_search.measuredHeight / 2
            val width = centerX * 2
            val height = app_search_view.measuredHeight
            radius = (sqrt(width * width + height * height.toDouble()) / 2).toFloat()
        }
        app_ibtn_back.setOnClickListener { closeSearchView() }
        app_ibtn_search.setOnClickListener { openSearchView() }
        app_ibtn_inverse.setOnClickListener {
            if (KeyboardUtils.isSoftInputVisible(this)) {
                KeyboardUtils.hideSoftInput(this)
            }
            app_ibtn_inverse.isSelected = !app_ibtn_inverse.isSelected
            adapter?.setInverse(app_ibtn_inverse.isSelected)
            adapter?.filter?.filter(app_et_search.text)
        }
        app_ibtn_clear.setOnClickListener {
            app_et_search.setText("")
            if (!KeyboardUtils.isSoftInputVisible(this)) {
                KeyboardUtils.showSoftInput(this)
            }
        }
        app_et_search.addTextChangedListener(object : TextWatcherImpl() {
            override fun onTextChanged(c: CharSequence, i: Int, i1: Int, i2: Int) {
                app_ibtn_inverse.visibility = if (c.isEmpty()) View.GONE else View.VISIBLE
                app_ibtn_clear.visibility = if (c.isEmpty()) View.GONE else View.VISIBLE
                adapter?.filter?.filter(c)
            }
        })
        app_ibtn_select_all.setOnClickListener {
            it.isSelected = !it.isSelected
            adapter?.selectAll(it.isSelected)
        }
        app_btn_import_count.setOnClickListener(object : OnClickListenerImpl() {
            override fun onViewClick(v: View, interval: Long) {
                if (KeyboardUtils.isSoftInputVisible(this@ScanActivity)) {
                    KeyboardUtils.hideSoftInput(this@ScanActivity)
                }
                if (selectList.isEmpty()) {
                    UiManager.showShort(R.string.app_have_not_select)
                } else {
                    importDialog.show()
                }
            }
        })
        app_ibtn_select_all.isEnabled = false // XML 设置无效，只能这里初始化
        val lm: RecyclerView.LayoutManager = LinearLayoutManager(this)
        app_rv_select.layoutManager = lm
        adapter = ViewAllAdapter(fileList, imported, object : ViewAllAdapter.Callback {
            override fun onDirTap(dirPath: String) { // empty impl
            }

            @SuppressLint("SetTextI18n")
            override fun onSelectResult(pathList: List<String>, total: Int) {
                LogUtils.e(pathList.size)
                if (total != 0) {
                    app_ibtn_select_all.isSelected = pathList.size == total
                }
                app_btn_import_count.text = getString(R.string.app_import_count, pathList.size)
                selectList.clear()
                selectList.addAll(pathList)
            }
        })
        adapter?.registerAdapterDataObserver(dataObserver)
        app_rv_select.adapter = adapter
        scanDialog.show()
        Observable.create(ObservableOnSubscribe { emitter: ObservableEmitter<Double?> -> emitter.onNext(traverseFile()) })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
//                .`as`<ObservableSubscribeProxy<Double>>(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(this)))
                .subscribe({
                    if (btnStopScan?.isSelected == false) {
                        scanDialog.dismiss()
                        updateUI()
                    }
                }) { throwable: Throwable -> LogUtils.e(throwable.message) }
    }

    private fun traverseFile(): Double {
        val start = System.currentTimeMillis()
        val result = SDCardUtils.getSDCardInfo()
        for (info in result) {
            if (!stopScan && "mounted" == info.state) {
                traverse(File(info.path))
            }
        }
        while (true) {
            val temp = scanCount
            SystemClock.sleep(1000)
            if (temp == scanCount) {
                stopScan = true
                val end = System.currentTimeMillis()
                val cost = (end - start - 1000).toDouble() / 1000
                LogUtils.e("总共耗时：$cost 秒")
                return cost
            }
        }
    }

    //    private int refreshLayoutFlag = 2000;
    @SuppressLint("SetTextI18n")
    @Synchronized
    private fun traverse(file: File?) {
        if (stopScan) {
            return
        }
        threadPool?.execute {
            val fileList = listable.listFile(file?.absolutePath)
            for (f in fileList) {
                if (stopScan) {
                    return@execute
                }
                scanCount++
                runOnUiThread { tvScanCount?.text = getString(R.string.app_already_scan, scanCount) }
                if (f.isFile) {
                    this.fileList.add(f)
                    pdfCount++
                    runOnUiThread {
                        toolbar?.title = getString(R.string.app_scan_result, pdfCount)
                        tvPdfCount?.text = getString(R.string.app_find, pdfCount)
                    }
                } else {
                    traverse(f)
                }
            }
        }
    }

    private fun initToolbar() {
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayShowTitleEnabled(false)
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setHomeAsUpIndicator(R.drawable.app_ic_action_back_black)
        }
        toolbar?.title = getString(R.string.app_scan_result, 0)
    }

    private fun openSearchView() {
        val animator = ViewAnimationUtils.createCircularReveal(app_search_view, centerX, centerY, 0f, radius)
        animator.duration = duration.toLong()
        animator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(animation: Animator) {
                app_search_view.visibility = View.VISIBLE
            }

            override fun onAnimationEnd(animation: Animator) {
                app_et_search.requestFocus()
                KeyboardUtils.showSoftInput(this@ScanActivity)
            }
        })
        animator.start()
    }

    private fun closeSearchView() {
        val animator = ViewAnimationUtils.createCircularReveal(app_search_view, centerX, centerY, radius, 0f)
        animator.duration = duration.toLong()
        animator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                if (KeyboardUtils.isSoftInputVisible(this@ScanActivity)) {
                    KeyboardUtils.hideSoftInput(this@ScanActivity)
                }
                app_search_view.visibility = View.GONE
            }
        })
        animator.start()
    }

    private fun setResultBack(type: Int, groupName: String?) {
        val data = Intent()
        data.putStringArrayListExtra(SelectActivity.EXTRA_SELECTED, selectList as ArrayList<String>)
        data.putExtra(SelectActivity.EXTRA_TYPE, type)
        data.putExtra(SelectActivity.EXTRA_GROUP_NAME, groupName)
        setResult(Activity.RESULT_OK, data)
        finish()
    }

    private fun updateUI() {
        val lac = AnimationUtils.loadLayoutAnimation(this, R.anim.app_layout_fall_down)
        app_rv_select.layoutAnimation = lac
        app_rv_select.adapter!!.notifyDataSetChanged()
        app_rv_select.scheduleLayoutAnimation()
    }

    companion object {
        const val EXTRA_IMPORTED = "EXTRA_IMPORTED"
        fun start(activity: Activity?, imported: ArrayList<String?>?, requestCode: Int) {
            val starter = Intent(activity, ScanActivity::class.java)
            starter.putStringArrayListExtra(EXTRA_IMPORTED, imported)
            activity!!.startActivityForResult(starter, requestCode)
        }
    }
}